src/framework/components/rigid-body/system.js

```diff
@@ -319,6 +319,53 @@ Object.assign(pc, function () {
             return result;
         },
 
+        /**
+         * @function
+         * @name pc.RigidBodyComponentSystem#raycastAll
+         * @description Raycast the world and return all entities the ray hits. It returns an array
+         * of {@link pc.RaycastResult}, one for each hit. If no hits are detected, the returned
+         * array will be of length 0.
+         * @param {pc.Vec3} start - The world space point where the ray starts.
+         * @param {pc.Vec3} end - The world space point where the ray ends.
+         * @returns {pc.RaycastResult[]} An array of raycast hit results (0 length if there were no hits).
+         */
+        raycastAll: function (start, end) {
+            var results = [];
```

Comment:
What if it returns null instead of a new array ? Presumably raycastAll might be called every frame and that might have an unnecessary impact on performance? Also the same with the AllHitsRayResultCallback below, can we reuse it?

Reply:
I deliberately didn't have it return `null` if no hits are detected. The reason is that it simplifies app-side code. Otherwise you have an `if` statement checking for two cases. Yes, calling the function will allocate an array per call, but under normal circumstances, the function will be returning results anyway. I don't think a single array represents much in the way of garbage. It's major GC stalls you want to avoid.

As for the callback, I don't think you can reuse it. I looked at the C++ code and it seems that the object accumulates all of the hit result info. There's not way to clear it after you've used it. Besides, `raycastFirst` is doing exactly what I'm doing now in `raycastAll`. I think it's OK.